src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -1405,6 +1428,141 @@ private void WritePackageDependencies()
                 }
             }
 
+            // This implementation was taken from PreprocessPackageDependenciesDesignTime.ExecuteCore
+            // with some minor modifications to use WriteItem instead of saving it into an array.
+            private void WritePackageDependenciesDesignTime()
+            {
+                List<string> packageDependencies = GetPackageDependencies();
+
+                var implicitPackageReferences = GetImplicitPackageReferences(_task.DefaultImplicitPackages);
+
+                // We have two types of data:
+                //
+                // 1) "PackageDependencies" which place a package in a given target/hierarchy
+                // 2) "PackageDefinitions" which provide general metadata about a package
+                //
+                // First, we scan PackageDependencies to build the set of packages in our target.
+
+                var allowItemSpecs = packageDependencies.ToHashSet(StringComparer.OrdinalIgnoreCase);
+
+                // Second, find PackageDefinitions that match our allowed item specs
+
+                foreach (var package in _lockFile.Libraries)
+                {
+                    if (string.IsNullOrEmpty(package.Name) || !allowItemSpecs.Contains(package.Name))
+                    {
+                        // Name is required
+                        // We are not interested in this definition (not top-level, or wrong target)
+                        continue;
+                    }
+
+                    var dependencyType = GetDependencyType(package.Type);
+
+                    if (dependencyType == DependencyType.Package ||
+                        dependencyType == DependencyType.Unresolved)
+                    {
+                        WriteItem(package.Name);
```

Comment:
I think the itemSpec needs to include the version, in the form returned by `GetPackageId`.

Reply:
fixed